File: net.akehurst.language/processor/src/test/java/net/akehurst/language/processor/vistraq/test_VistraqQuery.java
package net.akehurst.language.processor.vistraq;

import java.io.InputStreamReader;
import java.util.ArrayList;
import java.util.Collection;
import java.util.List;

import org.junit.Assert;
import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.junit.runners.Parameterized;
import org.junit.runners.Parameterized.Parameter;
import org.junit.runners.Parameterized.Parameters;

import net.akehurst.language.core.analyser.UnableToAnalyseExeception;
import net.akehurst.language.core.grammar.IGrammar;
import net.akehurst.language.core.grammar.RuleNotFoundException;
import net.akehurst.language.core.parser.ParseFailedException;
import net.akehurst.language.core.parser.ParseTreeException;
import net.akehurst.language.core.sppt.ISharedPackedParseTree;
import net.akehurst.language.processor.LanguageProcessor;
import net.akehurst.language.processor.OGLanguageProcessor;

@RunWith(Parameterized.class)
public class test_VistraqQuery {

    static class Data {
        public Data(final String grammarRule, final String queryStr) {
            this.grammarRule = grammarRule;
            this.queryStr = queryStr;
        }

        public final String grammarRule;
        public final String queryStr;

        // --- Object ---
        @Override
        public String toString() {
            return this.grammarRule + " : " + this.queryStr;
        }
    }

    @Parameters(name = "{0}")
    public static Collection<Object[]> data() {
        final List<Object[]> col = new ArrayList<>();
        col.add(new Object[] { new Data("linkSelector", "LINKED VIA * TO *") });
        col.add(new Object[] { new Data("artefactSelector", "*") });
        col.add(new Object[] { new Data("artefactSelector", "Requirement") });
        col.add(new Object[] { new Data("artefactSelector", "Requirement AS r") });
        col.add(new Object[] { new Data("pathExpression", "Requirement LINKED VIA * TO *") });
        col.add(new Object[] { new Data("pathExpression", "* LINKED VIA * TO *") });
        col.add(new Object[] { new Data("returnDefinition", "RETURN COLUMN Req CONTAINING r") });
        col.add(new Object[] { new Data("columnDefinition", "COLUMN Req CONTAINING r") });
        col.add(new Object[] { new Data("pathQuery", "MATCH Requirement AS r WHERE true") });
        col.add(new Object[] { new Data("query", "MATCH *") });
        col.add(new Object[] { new Data("query", "MATCH * AS n") });
        col.add(new Object[] { new Data("query", "MATCH Requirement AS r") });
        col.add(new Object[] { new Data("query", "MATCH Requirement AS r RETURN COLUMN Req CONTAINING r") });
        col.add(new Object[] { new Data("query", "MATCH Requirement AS r RETURN COLUMN Id CONTAINING r.identity") });
        col.add(new Object[] { new Data("expression", "r.type.name") });
        col.add(new Object[] { new Data("columnDefinition", "COLUMN Type CONTAINING r.type.name") });
        col.add(new Object[] { new Data("returnDefinition", "RETURN COLUMN Type CONTAINING r.type.name") });
        col.add(new Object[] { new Data("query", "MATCH Requirement AS r RETURN COLUMN Id CONTAINING r.identity COLUMN Type CONTAINING r.type.name") });
        col.add(new Object[] { new Data("query", "MATCH * LINKED VIA * TO *") });
        col.add(new Object[] { new Data("query", "MATCH * AS r LINKED VIA * TO * AS t") });
        col.add(new Object[] { new Data("query", "MATCH Requirement LINKED VIA * TO *") });
        col.add(new Object[] { new Data("query", "MATCH Requirement AS r LINKED VIA * TO * AS t") });
        col.add(new Object[] { new Data("query", "MATCH Requirement LINKED VIA * TO Test") });
        col.add(new Object[] { new Data("query", "MATCH Requirement AS r LINKED VIA * TO Test AS t") });
        col.add(new Object[] { new Data("query", "MATCH Requirement AS r LINKED VIA testedBy TO Test AS t") });
        col.add(new Object[] { new Data("query", "MATCH Requirement AS r LINKED VIA testedBy TO Test AS t") });
        col.add(new Object[] { new Data("query", "MATCH * LINKED VIA * TO * LINKED VIA * TO *") });
        col.add(new Object[] { new Data("query", "MATCH * AS n RETURN COLUMN count CONTAINING COUNT n") });
        col.add(new Object[] { new Data("query", "MATCH * NOT (LINKED VIA * TO *)") });
        col.add(new Object[] { new Data("query", "MATCH * NOT (LINKED VIA * FROM *)") });
        col.add(new Object[] { new Data("query", "MATCH * ( LINKED VIA * TO * AND LINKED VIA * TO *  )") });
        col.add(new Object[] { new Data("query", "FROM query1") });
        col.add(new Object[] { new Data("query", "FROM query1 RETURN COLUMN Req CONTAINING r") });
        col.add(new Object[] { new Data("query", "FROM qs1.query1 RETURN COLUMN Req CONTAINING r") });
        col.add(new Object[] { new Data("expression", "a") });
        col.add(new Object[] { new Data("expression", "a.prop") });
        col.add(new Object[] { new Data("expression", "a.meth()") });
        col.add(new Object[] { new Data("expression", "a.prop + 1") });
        col.add(new Object[] { new Data("expression", "a.prop == a.prop") });
        return col;
    }

    @Parameter
    public Data data;

    private LanguageProcessor processor;

    private String clean(final String str) {
        String res = str.replaceAll(System.lineSeparator(), " ");
        res = res.trim();
        return res;
    }

    @Before
    public void setup() {
        try {
            final OGLanguageProcessor oglProc = new OGLanguageProcessor();
            final InputStreamReader reader = new InputStreamReader(test_VistraqQuery.class.getResourceAsStream("/vistraq/Query.ogl"));
            final IGrammar grammar = oglProc.process(reader, "grammarDefinition", IGrammar.class);

            this.processor = new LanguageProcessor(grammar, null);
        } catch (final ParseFailedException e) {
            Assert.fail("Parsing query grammar failed " + e.getLongestMatch());
        } catch (final UnableToAnalyseExeception e) {
            Assert.fail("Analysing query grammar failed");
        }
    }

    @Test
    public void test() throws ParseFailedException, ParseTreeException, RuleNotFoundException {

        final String queryStr = this.data.queryStr;
        final String grammarRule = this.data.grammarRule;
        final ISharedPackedParseTree tree = this.processor.getParser().parse(grammarRule, queryStr);
        Assert.assertNotNull(tree);
        final String resultStr = this.clean(tree.asString());
        Assert.assertEquals(queryStr, resultStr);

    }

}

File: net.akehurst.language/processor/src/jvm8Test/kotlin/processor/vistraq/test_VistraqQuery.kt
package net.akehurst.language.processor.vistraq

import java.io.BufferedReader
import java.io.IOException
import java.io.InputStream
import java.io.InputStreamReader
import java.util.ArrayList

import org.junit.Assert
import org.junit.Before
import org.junit.Test
import org.junit.runner.RunWith
import org.junit.runners.Parameterized
import org.junit.runners.Parameterized.Parameters

import net.akehurst.language.api.processor.LanguageProcessor
import net.akehurst.language.api.sppt.SharedPackedParseTree
import net.akehurst.language.processor.processor

@RunWith(Parameterized::class)
class test_QueryParserValid(val data:Data) {

    companion object {

        private val grammarStr = """
namespace com.itemis.typedgraph.query

grammar Query {

    skip WHITE_SPACE = "\s+" ;
	skip SINGLE_LINE_COMMENT = "/\*[^*]*\*+(?:[^*/][^*]*\*+)*/" ;
	skip MULTI_LINE_COMMENT = "//.*?$" ;

	query = singleQuery | unionedQuery ;
	singleQuery = timespanDefinition? querySource? returnDefinition? ;
	unionedQuery = query unionOperator query ;
	unionOperator = 'UNION' ; //TODO: handle distinct or duplicates

    timespanDefinition = 'FOR' timeDef nameDefinition? ;
	timeDef =  time | timespan ;
	time = 'TIME' timePoint ;
	timePoint = 'start' | 'now' | timestamp ;
	timestamp = SINGLE_QUOTE_STRING ;
	timespan = 'TIMESPAN' timeRange 'EVERY' period ;
	timeRange = 'all' | timePoint 'UNTIL' timePoint ;
	period = 'second' | 'minute' | 'hour' | 'day' | 'week' | 'month' | 'year' ;


	querySource = pathQuery | storedQueryReference ;
	storedQueryReference = 'FROM' STORED_QUERY_ID ;
    pathQuery = 'MATCH' pathExpression whereClause? ;
    pathExpression =  nodeSelector linkedNodeSelectorPath? ;

	nodeSelector = nodeTypeReferenceExpression nameDefinition? ;
    nodeTypeReferenceExpression = ANY_NODE_TYPE | NONE_NODE_TYPE | nodeTypeReference | negatedNodeTypeReferenceExpression | nodeTypeReferenceGroup ;
    negatedNodeTypeReferenceExpression = 'NOT' nodeTypeReferenceExpression ;
    nodeTypeReferenceGroup = '(' [nodeTypeReferenceExpression / 'OR' ]+ ')' ;

    linkedNodeSelector = linkSelectorExpression ;
    linkedNodeSelectorNegated = linkSelectorNegated ;
	linkSelectorExpression = linkSelector | linkSelectorGroupedPath ;
	linkSelector = 'LINKED' links? multiplicity? via? ('TO'|'FROM'|'WITH') nodeSelector ;

	links = 'USING' POSITIVE_INT '..' UNLIMITED_POSITIVE_INT 'LINKS' ;
	multiplicity = POSITIVE_INT '..' UNLIMITED_POSITIVE_INT 'TIMES' ;

	via = 'VIA'  linkTypeReferenceExpression ;
	linkSelectorNegated = 'NOT' linkSelectorGroupedPath ;
	linkSelectorGroupedPath = '(' linkSelectorGroupedItem ')' ;
	linkSelectorGroupedItem = linkedNodeSelectorPath | linkSelectorOperator ;
	//linkSelectorGroupedOperator = '(' linkSelectorOperator ')';
	linkSelectorOperator = linkedNodeSelectorPath logicalOperator linkedNodeSelectorPath ;

	linkedNodeSelectorPath = linkedNodeSelectorPathNormal | linkedNodeSelectorPathNegated ;
	linkedNodeSelectorPathNormal = linkedNodeSelector+ ;
	linkedNodeSelectorPathNegated = linkedNodeSelector*  linkedNodeSelectorNegated ;

    linkTypeReferenceExpression = ANY_LINK_TYPE | linkTypeReference | negatedLinkTypeReferenceExpression | linkTypeReferenceGroup ;
    negatedLinkTypeReferenceExpression = 'NOT' linkTypeReferenceExpression ;
    linkTypeReferenceGroup = '(' [linkTypeReferenceExpression / 'OR' ]+ ')' ;

	whereClause = 'WHERE' expression ;

	returnDefinition = returnExpression | returnTable | returnSelect | returnGraph ;

	returnExpression = 'RETURN' aggregateFunctionCallOrExpression ;

	returnTable = 'RETURN' 'TABLE' columnDefinition+ orderBy? ;
	columnDefinition = 'COLUMN' NAME 'CONTAINING' aggregateFunctionCallOrExpression ;

    returnSelect = 'RETURN' 'SELECT' selectList;
    selectList = [selectItem / ',']* ;
    selectItem = NAME | propertyReference ;
    propertyReference = NAME '.' NAME ;

	returnGraph = 'RETURN' 'GRAPH' subgraphConstruction ;
	subgraphConstruction = nodeConstruction linkedSubGraphConstruction? whereConstructionClause?;
	nodeConstruction = nodeTypeConstructionExpression '{' nodeIdentityExpressionConstruction nodePropertyAssignmentExpressionList? '}' nameDefinition?;
	nodeTypeConstructionExpression = nodeTypeReference < expression ;
    nodeIdentityExpressionConstruction = expression ;
    nodePropertyAssignmentExpressionList = '|' nodePropertyAssignmentExpression+ ;
    nodePropertyAssignmentExpression = NAME ':=' expression ;

	linkedSubGraphConstruction = linkedNodeConstructionExpression+ ;
	linkedNodeConstructionExpression = linkedNodeConstructionPath | linkedNodeConstructionGroup ;
	linkedNodeConstructionPath = linkedNodeConstruction+ ;
	linkedNodeConstruction = 'LINKED' 'VIA' linkTypeReference ('TO'|'FROM') nodeConstruction ;
	linkedNodeConstructionGroup = '(' linkedNodeConstructionGroupItem ')' ;
	linkedNodeConstructionGroupItem = linkedNodeConstructionPath |  linkedNodeConstructionGroupItemOperator ;
	linkedNodeConstructionGroupItemOperator =  linkedNodeConstructionExpression 'AND' linkedNodeConstructionExpression ;

	whereConstructionClause = 'WHERE' constructionExpression ;

	constructionExpression = andConstructionExpression
	                       < propertyAssignment
                           ;

	propertyAssignment = propertyCall ':=' expression ;
	andConstructionExpression = constructionExpression 'AND' constructionExpression ;

	aggregateFunctionCallOrExpression = aggregateFunctionCall | expression ;
	aggregateFunctionCall = aggregateFunctionName '(' expression ')' ;
    expression	= groupExpression
				< conditionalExpression
				< infixFunction
				< literalValue
				< root
				< propertyCall
				< methodCall
            ;

    nameDefinition = 'AS' NAME ;

    groupExpression = '(' expression ')'  ;
	root = NAME;
	propertyCall = expression '.' NAME ;
	methodCall = expression '.' NAME '(' argList ')';
	argList = [expression / ',']* ;
	infixFunction = expression operator expression ;
	operator = arithmeticOperator | comparisonOperator | logicalOperator ;
	arithmeticOperator = '+' | '-' | '*' | '/' ;
    comparisonOperator = '==' | '!=' | '<' | '>' | '<=' | '>=' ;
    logicalOperator= 'AND' | 'OR' | 'XOR' ;
    conditionalExpression = expression '?' expression ':' expression ;


	orderBy = 'ORDER' 'BY' [columnOrder / ',']+ ;
	columnOrder = NAME ('ASCENDING' | 'DESCENDING')? ;

	ANY_NODE_TYPE = '*' ;
	NONE_NODE_TYPE = 'none' ;
	nodeTypeReference = NAME ;
	ANY_LINK_TYPE = '*' ;
	linkTypeReference = NAME ;
	aggregateFunctionName = NAME ;
	STORED_QUERY_ID = "([a-zA-Z_][a-zA-Z0-9_]*)([.][a-zA-Z_][a-zA-Z0-9_]*)?" ; // project metricDef OR metricSet.metricDef
	NAME = "[a-zA-Z_][a-zA-Z0-9_]*" ;
	POSITIVE_INT = "[0-9]+" ;
	UNLIMITED_POSITIVE_INT = "[0-9]+" | '*' ;

    literalValue = BOOLEAN | SINGLE_QUOTE_STRING | INTEGER | REAL | NULL ;
    BOOLEAN = 'true' | 'false' ;
    SINGLE_QUOTE_STRING = "'(?:\\?.)*?'" ;
    INTEGER = "[0-9]+" ;
    REAL = "[0-9]+[.][0-9]+" ;
    NULL = 'null' ;
}
        """.trimIndent()
        var processor: LanguageProcessor = tgqlprocessor()

        var sourceFiles = arrayOf("vistraq/sampleValidQueries.txt")

        fun tgqlprocessor() : LanguageProcessor {
            //val grammarStr = ClassLoader.getSystemClassLoader().getResource("vistraq/Query.ogl").readText()
            return processor(grammarStr)
         }

        @JvmStatic
        @Parameters(name = "{0}")
        fun data(): Collection<Array<Any>> {
            val col = ArrayList<Array<Any>>()
            for (sourceFile in test_QueryParserValid.sourceFiles) {
                val `is` = ClassLoader.getSystemClassLoader().getResourceAsStream(sourceFile)
                val br = BufferedReader(InputStreamReader(`is`))
                var line: String? = br.readLine()
                while (null != line) {
                    line = line.trim { it <= ' ' }
                    if (line.isEmpty()) {
                        // blank line
                        line = br.readLine()
                    } else if (line.startsWith("//")) {
                        // comment
                        line = br.readLine()
                    } else {
                        col.add(arrayOf(Data(sourceFile, line)))
                        line = br.readLine()
                    }
                }
            }
            return col
        }
    }

    class Data(val sourceFile: String, val queryStr: String) {

        // --- Object ---
        override fun toString(): String {
            return this.sourceFile + ": " + this.queryStr
        }
    }

    @Test
    fun test() {
        val queryStr = this.data.queryStr
        val result = processor.parse("query", queryStr)
        Assert.assertNotNull(result)
        val resultStr = result.asString
        Assert.assertEquals(queryStr, resultStr)
    }



}
